qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java

```diff
@@ -363,4 +352,96 @@ public void testRecoveryWithSoftDeletes() throws Exception {
+            createIndex(index, settings.build());
+            updateIndexSettings(index, Settings.builder().put("index.routing.allocation.include._name", "node-1,node-2"));
+            ensureGreen(index);
+            // move the primary from node-1 to upgraded-node-0 so we can index documents with sequence numbers
+            // and prevent allocating the replica on upgraded-node-2 until we trim translog on the primary so recovery won't replay ops
+            updateIndexSettings(index, Settings.builder().put("index.routing.allocation.include._name", "upgraded-node-0,node-2"));
+            ensureGreen(index);
+            indexDocs(index, 0, 10);
+            syncedFlush(index);
+        } else if (CLUSTER_TYPE == ClusterType.UPGRADED) {
+            updateIndexSettings(index, Settings.builder()
+                .put(IndexSettings.INDEX_TRANSLOG_RETENTION_AGE_SETTING.getKey(), "-1")
+                .put(IndexSettings.INDEX_TRANSLOG_RETENTION_SIZE_SETTING.getKey(), "-1")
+                .put(IndexSettings.INDEX_TRANSLOG_GENERATION_THRESHOLD_SIZE_SETTING.getKey(), "256b"));
+            // index more documents to roll translog so we can trim translog
+            indexDocs(index, 10, 100);
+            assertBusy(() -> {
+                Map<String, ?> stats = entityAsMap(client().performRequest(new Request("GET", index + "/_stats?level=shards")));
+                Integer translogOps = (Integer) XContentMapValues.extractValue("_all.primaries.translog.operations", stats);
+                assertThat(translogOps, equalTo(100));
+            });
+            updateIndexSettings(index, Settings.builder().put("index.routing.allocation.include._name", "upgraded-node-0,upgraded-node-2"));
+            ensureGreen(index);
+        }
+    }
+
+    public void testRollingUpgradeWithSyncedFlush() throws Exception {
+        final String index = "rolling_upgrade_with_synced_flush";
+        if (CLUSTER_TYPE == ClusterType.OLD) {
+            createIndex(index, Settings.builder()
+                .put(INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), "30s")
+                .put(IndexMetaData.INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), 1)
+                .put(IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), randomIntBetween(1, 2)).build());
+            ensureGreen(index);
+            indexDocs(index, 0, 40);
+            syncedFlush(index);
+        } else if (CLUSTER_TYPE == ClusterType.MIXED) {
+            ensureGreen(index);
+            assertNoFileBasedRecovery(index);
+            indexDocs(index, 40, 50);
+            syncedFlush(index);
+        } else {
+            ensureGreen(index);
+            assertNoFileBasedRecovery(index);
+            indexDocs(index, 90, 60);
+            syncedFlush(index);
+        }
+    }
+
+    private void syncedFlush(String index) throws Exception {
+        // We have to spin synced-flush requests here because we fire the global checkpoint sync for the last write operation.
```

Comment:
I think there is a very small chance of a race condition here. If both the synced flush and node stop runs before the global checkpoint sync, the node will come up with an older safe commit and will revert to file based recovery. I could be wrong of course. The likelihood seems very small though. I did not find a good way to check if the global checkpoint sync has started/completed, maybe you have an idea?

Reply:
Well spotted. I have an idea and will try it out.